File: src/main/java/com/github/rafaelldi/tyeplugin/run/TyeSettingsEditor.java
package com.github.rafaelldi.tyeplugin.run;

import com.intellij.openapi.fileChooser.FileChooserDescriptor;
import com.intellij.openapi.options.SettingsEditor;
import com.intellij.openapi.ui.TextBrowseFolderListener;
import com.intellij.openapi.ui.TextFieldWithBrowseButton;
import com.intellij.openapi.vfs.LocalFileSystem;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.ui.JBIntSpinner;
import com.intellij.ui.components.JBCheckBox;
import com.intellij.ui.components.JBRadioButton;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;

import static com.github.rafaelldi.tyeplugin.run.OptionsConstants.*;

public class TyeSettingsEditor extends SettingsEditor<TyeRunConfiguration> {
    private JPanel panel;
    private TextFieldWithBrowseButton pathField;
    private JBIntSpinner portSpinner;
    private JBCheckBox noBuildCheckBox;
    private JBCheckBox dockerCheckBox;
    private JBCheckBox dashboardCheckBox;
    private JBRadioButton debugRadioButton;
    private JBRadioButton infoRadioButton;
    private JBRadioButton quietRadioButton;

    @Override
    protected void resetEditorFrom(@NotNull TyeRunConfiguration runConfig) {
        VirtualFile pathArgument = runConfig.getPathArgument();
        if (pathArgument != null){
            pathField.setText(pathArgument.getPath());
        } else {
            pathField.setText("");
        }
        portSpinner.setNumber(runConfig.getPortArgument());
        noBuildCheckBox.setSelected(runConfig.getNoBuildArgument());
        dockerCheckBox.setSelected(runConfig.getDockerArgument());
        dashboardCheckBox.setSelected(runConfig.getDashboardArgument());

        String verbosity = runConfig.getVerbosityArgument();
        if (verbosity.equals(DEBUG_VERBOSITY)) {
            debugRadioButton.setSelected(true);
        } else if (verbosity.equals(INFO_VERBOSITY)) {
            infoRadioButton.setSelected(true);
        } else {
            quietRadioButton.setSelected(true);
        }
    }

    @Override
    protected void applyEditorTo(@NotNull TyeRunConfiguration runConfig) {
        runConfig.setPathArgument(LocalFileSystem.getInstance().findFileByPath(pathField.getText()));
        runConfig.setPortArgument(portSpinner.getNumber());
        runConfig.setNoBuildArgument(noBuildCheckBox.isSelected());
        runConfig.setDockerArgument(dockerCheckBox.isSelected());
        runConfig.setDashboardArgument(dashboardCheckBox.isSelected());

        if (debugRadioButton.isSelected()) {
            runConfig.setVerbosityArgument(DEBUG_VERBOSITY);
        } else if (infoRadioButton.isSelected()) {
            runConfig.setVerbosityArgument(INFO_VERBOSITY);
        } else {
            runConfig.setVerbosityArgument(QUIET_VERBOSITY);
        }
    }

    @Override
    protected @NotNull JComponent createEditor() {
        return panel;
    }

    private void createUIComponents() {
        pathField = new TextFieldWithBrowseButton();
        FileChooserDescriptor descriptor = new FileChooserDescriptor(true, true, false, false, false, false);
        descriptor.withFileFilter(vf ->
                vf.getName().equalsIgnoreCase("tye.yaml") ||
                        (vf.getExtension() != null &&
                                (vf.getExtension().equalsIgnoreCase("sln") ||
                                        vf.getExtension().equalsIgnoreCase("csproj") ||
                                        vf.getExtension().equalsIgnoreCase("fsproj"))));
        pathField.addBrowseFolderListener(new TextBrowseFolderListener(descriptor));

        portSpinner = new JBIntSpinner(DEFAULT_PORT, 0, 65353);
    }
}

File: src/main/kotlin/com/github/rafaelldi/tyeplugin/run/TyeSettingsEditor.kt
package com.github.rafaelldi.tyeplugin.run

import com.intellij.openapi.fileChooser.FileChooserDescriptor
import com.intellij.openapi.options.SettingsEditor
import com.intellij.openapi.ui.LabeledComponent
import com.intellij.openapi.ui.TextBrowseFolderListener
import com.intellij.openapi.ui.TextFieldWithBrowseButton
import com.intellij.openapi.ui.VerticalFlowLayout
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.ui.JBIntSpinner
import com.intellij.ui.TitledSeparator
import com.intellij.ui.components.JBCheckBox
import java.awt.BorderLayout
import javax.swing.JComponent
import javax.swing.JPanel

class TyeSettingsEditor : SettingsEditor<TyeRunConfiguration>() {
    companion object {
        const val DEFAULT_PORT: Int = 8000
        const val MIN_PORT: Int = 0
        const val MAX_PORT: Int = 65353
    }

    private lateinit var panel: JPanel
    private lateinit var pathField: LabeledComponent<TextFieldWithBrowseButton>
    private lateinit var portField: LabeledComponent<JBIntSpinner>
    private lateinit var noBuildOption: JBCheckBox
    private lateinit var dockerOption: JBCheckBox
    private lateinit var dashboardOption: JBCheckBox

    override fun createEditor(): JComponent {
        createUIComponents()
        return panel
    }

    override fun resetEditorFrom(runConfig: TyeRunConfiguration) {
        pathField.component.text = runConfig.pathArgument?.path ?: ""
        portField.component.number = runConfig.portArgument
        noBuildOption.isSelected = runConfig.noBuildArgument
        dockerOption.isSelected = runConfig.dockerArgument
        dashboardOption.isSelected = runConfig.dashboardArgument
    }

    override fun applyEditorTo(runConfig: TyeRunConfiguration) {
        runConfig.pathArgument = LocalFileSystem.getInstance().findFileByPath(pathField.component.text)
        runConfig.portArgument = portField.component.number
        runConfig.noBuildArgument = noBuildOption.isSelected
        runConfig.dockerArgument = dockerOption.isSelected
        runConfig.dashboardArgument = dashboardOption.isSelected
    }

    private fun createUIComponents() {
        panel = JPanel().apply {
            layout = VerticalFlowLayout(VerticalFlowLayout.TOP)

            val pathTextField = TextFieldWithBrowseButton().apply {
                addBrowseFolderListener(
                    TextBrowseFolderListener(
                        FileChooserDescriptor(true, true, false, false, false, false)
                            .withFileFilter { vf ->
                                vf.name.toLowerCase() == "tye.yaml" ||
                                    vf.extension?.toLowerCase() == "sln" ||
                                    vf.extension?.toLowerCase() == "csproj" ||
                                    vf.extension?.toLowerCase() == "fsproj"
                            }
                            .withTitle("Select Path Argument")
                    )
                )
            }
            pathField = LabeledComponent.create(pathTextField, "Path")
            pathField.labelLocation = BorderLayout.WEST
            add(pathField)

            add(TitledSeparator("Options"))

            val portSpinner = JBIntSpinner(DEFAULT_PORT, MIN_PORT, MAX_PORT)
            portField = LabeledComponent.create(portSpinner, "Dashboard port")
            portField.labelLocation = BorderLayout.WEST
            add(portField)

            noBuildOption = JBCheckBox("--no-build, does not build projects before running")
            add(noBuildOption)

            dockerOption = JBCheckBox("--docker, run projects as docker containers")
            add(dockerOption)

            dashboardOption = JBCheckBox("--dashboard, launch dashboard on run")
            add(dashboardOption)
        }
    }
}
